cpp/src/arrow/extension/tensor_array.cc

```diff
@@ -0,0 +1,161 @@
+
+  rj::Value shape(rj::kArrayType);
+  for (auto v : shape_) {
+    shape.PushBack(v, allocator);
+  }
+  document.AddMember(rj::Value("shape", allocator), shape, allocator);
+  document.AddMember(rj::Value("order", allocator),
+                     rj::Value(is_row_major_ ? "C" : "F", allocator), allocator);
+  if (!document.HasMember("metadata")) {
+    document.AddMember(rj::Value("metadata", allocator), rj::Value(metadata_, allocator),
+                       allocator);
+  }
+
+  rj::StringBuffer buffer;
+  rj::Writer<rj::StringBuffer> writer(buffer);
+  document.Accept(writer);
+  return buffer.GetString();
+}
+
+Result<std::shared_ptr<DataType>> TensorArrayType::Deserialize(
+    std::shared_ptr<DataType> storage_type, const std::string& serialized_data) const {
+  rj::Document document;
+  document.Parse(serialized_data);
+  if (!document.HasMember("shape") || !document.HasMember("order") ||
+      !document["shape"].IsArray()) {
+    return Status::Invalid("Invalid serialized data: ", serialized_data);
+  }
+
+  std::vector<int64_t> shape;
+  for (auto& v : document["shape"].GetArray()) {
+    shape.emplace_back(v.GetInt64());
+  }
+
+  bool is_row_major = document["order"].GetString() == std::string("C");
+
+  document.RemoveMember("shape");
+  document.RemoveMember("order");
+
+  rj::StringBuffer buffer;
+  rj::Writer<rj::StringBuffer> writer(buffer);
+  document.Accept(writer);
+  const std::string remaining_metadata = buffer.GetString();
+
+  return std::make_shared<TensorArrayType>(storage_type, shape, is_row_major,
+                                           remaining_metadata);
+}
+
+std::shared_ptr<Array> TensorArrayType::MakeArray(std::shared_ptr<ArrayData> data) const {
+  return std::make_shared<ExtensionArray>(data);
+}
+
+std::shared_ptr<TensorArray> TensorArrayType::MakeArray(
+    std::shared_ptr<Tensor> tensor) const {
+  ARROW_PREDICT_TRUE(tensor->is_row_major());
+
+  auto ext_type = tensor_array(tensor->type(), tensor->shape(), tensor->is_row_major());
+
+  auto data = ArrayData::Make(ext_type, tensor->size(), {nullptr, tensor->data()});
+  data->type = ext_type;
+  return std::make_shared<TensorArray>(data);
+}
+
+std::shared_ptr<Tensor> TensorArrayType::ToTensor(std::shared_ptr<Array> arr) const {
+  ARROW_DCHECK_EQ(arr->null_count(), 0) << "Null values not supported in tensors.";
+  std::vector<int64_t> shape = shape_;
+  shape.insert(shape.begin(), 1, arr->length());
+  return std::make_shared<Tensor>(storage_type(), arr->data()->buffers[1], shape,
+                                  strides());
+}
+
+std::shared_ptr<ExtensionType> tensor_array(const std::shared_ptr<DataType>& storage_type,
+                                            const std::vector<int64_t>& shape,
+                                            const bool& is_row_major,
+                                            const std::string& metadata) {
+  // TODO: Should we register the extension type here?
```

Comment:
I don't think it should be registered here (that would give unpredictable behaviour, as it would only get registered if you use this function, and eg not when reading IPC with such an extension type)

Reply:
Indeed. Still looking for alternatives.